numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -1374,6 +1384,47 @@ PyArray_PromoteTypes(PyArray_Descr *type1, PyArray_Descr *type2)
             break;
     }
 
+    if (PyDataType_HASFIELDS(type1) || PyDataType_HASFIELDS(type2)) {
+        if (!PyDataType_HASFIELDS(type1) || !PyDataType_HASFIELDS(type2)) {
+            PyErr_SetString(PyExc_TypeError, "invalid type promotion");
+            return NULL;
+        }
+
+        int compatible = 0;
+        if (type1->names == type2->names) {
+            compatible = 1;
+        }
+        else {
+            compatible = PyObject_RichCompareBool(type1->names, type2->names,
+                                                  Py_EQ);
+            if (compatible == -1) {
+                PyErr_Clear();
+                return NULL;
+            }
+        }
+
+        if (compatible) {
+            static PyObject *repack_func = NULL;
+            npy_cache_import("numpy.lib.recfunctions", "repack_fields",
+                             &repack_func);
+            if (repack_func == NULL) {
+                return NULL;
+            }
+            type1 = (PyArray_Descr*)PyObject_CallFunction(repack_func,
+                                                          "Oii", type1, 0, 1);
+            if (type1 == NULL) {
+                return NULL;
+            }
+            PyArray_Descr *ret = PyArray_DescrNewByteorder(type1, NPY_NATIVE);
+            Py_DECREF(type1);
+            return ret;
+        }
```

Comment:
This is currently incorrect for:
```python
In [2]: dt1 = np.dtype([("", "i4")])                                                                                    
In [3]: dt2 = np.dtype([("", "i8")])                                                                                    
In [6]: np.promote_types(dt1, dt2)                                                                                      
Out[6]: dtype([('f0', '<i4')])

In [7]: np.promote_types(dt2, dt1)                                                                                      
Out[7]: dtype([('f0', '<i8')])
```
Which has to either not work for now, or probably promote each included dtype recursively.

Reply:
Yes, I totally missed this possibility.